### ISSUE TITLE
md5 and sha module will be deprecated in python 2.6

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2763]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#2763|
|Type|enhancement|
|Created|2007-08-01 08:39:49Z|
|Branch|https://github.com/twisted/twisted/tree/hashlib-2763-3|

We have to use the hashlib module available since python 2.5.

Note that I saw this because stdio tests are failing on python trunk. I'm not sure it should be sensitive to warnings emitted...

Attachments:

* [fix-md5-and-sha.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bf5/bf5d4e7c8a1af61855b5227279a6744c3bde8a8a/f7bc3e35c2d3cf1007e3ea6bfe6ee26909d76e8c.patch) (52076 bytes) - added by devinus on 2008-02-18 22:24:52Z - 
* [hashlib.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bf5/bf5d4e7c8a1af61855b5227279a6744c3bde8a8a/882852da75a34e1d83e1ed63f7f6cb2f7dcd05e4.patch) (45353 bytes) - added by wsanchez on 2008-11-12 22:08:22Z - 
* [hashlib.v2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/bf5/bf5d4e7c8a1af61855b5227279a6744c3bde8a8a/51b82757cdf06fd63f6f705c61c91a2cf27bc84b.patch) (48374 bytes) - added by wsanchez on 2008-11-12 22:34:19Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2763 2763
type__enhancement enhancement
reporter__therve therve
priority__normal normal
milestone__Python_2_6 Python-2.6
branch__branches_hashlib_2763_3 branches/hashlib-2763-3
branch_author__wsanchez__exarkun wsanchez, exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1185957589000000 1185957589000000
changetime__1227124618000000 1227124618000000
version__None None
owner__ 
cc__ralphm cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

The stdio tests should definitely be changed to not break as a result of this deprecation.

**Comment by twisted-trac** [comment]
|<img alt="devinus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|devinus set owner to devinus<br>devinus set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="devinus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|devinus removed owner<br>devinus set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_7"></a>|@dreid set owner to devinus|
|-|-|

You should fall back to importing sha and md5 if the hashlib import fails

Something like
```
try:
    from hashlib import md5, sha1 as sha
except ImportError:
    from md5 import md5
    from sha import sha
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

<itamar> instead, why not have the `twisted.python.compat` module install a fake hashlib module on Python versions that don't provide it and then use `hashlib` everywhere?

**Comment by twisted-trac** [comment]
|<img alt="devinus's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|devinus set status to `assigned`|
|-|-|

```
try:
    import hashlib
except ImportError:
    import md5
    import sha


    class hashlib(object):
        """Mimic a small subset of the Python 2.5 hashlib module."""
        __module__ = 'hashlib'

        @classmethod
        def new(cls, name, string=''):
            if name in ('MD5', 'md5'):
                return md5.new(string)
            elif name in ('SHA1', 'sha1'):
                return sha.new(string)
            raise ValueError, "unsupported hash type"

        @classmethod
        def md5(cls, string=''):
            return md5.new(string)

        @classmethod
        def sha1(cls, string=''):
            return sha.new(string)


    sys.modules['hashlib'] = hashlib
```

I had all of the files using the old md5 and sha modules converted over, but some were failing the test suite, so I'll need more time.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

There's <http://code.krypto.org/python/hashlib/>.  I wonder if we can somehow make use of this.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_11"></a>|@wsanchez commented|
|-|-|

Attached hashlib.patch got rid of thw warnings for me and all tests passed.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set owner to @wsanchez<br>@exarkun set status to `new`|
|-|-|

```
22:14 < wsanchez> How would .compat install the fake module?  I can spend the time to clean up all the other files, just not sure what you had in mind
22:14 < wsanchez> We can do import twisted.python.hashlib which imports the right thing; that'd be easy
22:15 < exarkun> I think that would be fine.
22:15 < exarkun> Preferable to stuffing fakes into sys.modules, in fact.
22:15 < wsanchez> Or from twisted.python.compat import md5, sha1, but that seems a little less nice
22:15 < exarkun> Having twisted.python.hashlib.md5 and twisted.python.hashlib.sha1 sounds good.
22:16 < wsanchez> Yeah, I wouldn't want to monkey with sys.modules; that seems like looking for some kind of sneaky pain later
22:16 < wsanchez> OK I'll do that.
```

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_13"></a>|@wsanchez commented|
|-|-|

Also replacing 'sha1 as sha' imports to simply 'sha1'.  Makes searching for uses much easier and will make it easier to switch to directly using Python's hashlib in the future.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_15"></a>|@wsanchez commented|
|-|-|

Tests pass with the v2 patch on my OS X 10.5 box with Py2.5.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_16"></a>|@wsanchez commented|
|-|-|

(In [25340]) Branch for [#2763](https://github.com/twisted/twisted/issues/2763): md5 and sha module will be deprecated in python 2.6

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_17"></a>|@wsanchez set status to `closed`|
|-|-|

(In [25341]) Add a twisted.python.hashlib module which imports hashlib or (if that
fails) md5 and sha, which are deprecated in Python 2.6.
Fixes [#2763](https://github.com/twisted/twisted/issues/2763).

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set status to `reopened`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun set status to `new`|
|-|-|

There are some test failures on Windows.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_20"></a>|@exarkun commented|
|-|-|

Actually, I suppose that was premature.  There seem to be mad failures on _most_ platforms.  Maybe all.  And the test suite is taking **ages**.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_21"></a>|@wsanchez commented|
|-|-|

Failures appear limited to conch:
```
  File "c:\twistedbuildbot24\WXP32-full2.4-scmikes-select\Twisted\twisted\conch\ssh\transport.py", line 1226, in _getCipher
    counter=_Counter(iv, mod.block_size))
exceptions.TypeError: 'module' object is not callable
```
And so far only Windows has completed (Ubuntu lost the builder). It's going…very…slowly…

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_22"></a>|@wsanchez commented|
|-|-|

Whoops. r25343 should address that.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_23"></a>|@wsanchez commented|
|-|-|

OK, all tests are passing now:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/hashlib-2763

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_24"></a>|@therve commented|
|-|-|

Note that tests passing, but you're not using hashlib...
"import hashlib" in twisted.python.hashlib tries to make it import itself. You have to use `__import__`, or rename the module in twisted.python.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_25"></a>|@wsanchez set status to `assigned`|
|-|-|

OK:
```
__all__ = ["md5", "sha1"]

try:
    hashlib = __import__("hashlib")
    md5  = hashlib.md5
    sha1 = hashlib.sha1
except ImportError:
    from md5 import md5
    from sha import sha as sha1
```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_28"></a>|@exarkun commented|
|-|-|

  1. twisted/python/hashlib.py
    1. no copyright statement
    1. no test-case-name
    1. no module docstring
    1. no unit tests
    1. only the `__import__` should be protected by the try/except ImportError.  The two name bindings should happen outside.
    1. `__all__` is cool; `hashlib` should probably be `_hashlib` anyway, though, because sometimes people don't pay attention.
  1. twisted/conch/scripts/ckeygen.py
    1. copyright statement should be updated
    1. the module has some other random junk that might as well be cleaned up:
      1. module docstring is commented out
      1. there's still a cvs id marker which should be deleted
      1. imports appear in the wrong order
      1. `printFingerprint` uses deprecated apis
    1. the module has no unit test coverage whatsoever; any code being changed must have tests
  1. twisted/cred/util.py
    1. no test-case-name
    1. no unit tests
    1. actually, this module is entirely unused in twisted.  Should be deprecated, I think.
  1. twisted/words/protocols/oscar.py
    1. no unit tests
    1. copyright date should be updated
    1. old `__future__` import should be deleted
  1. twisted/words/protocols/msn.py
    1. update copyright date
    1. some trailing whitespace should be stripped
    1. `handle_CHL` is untested
  1. twisted/mail/maildir.py
    1. copyright date needs updating
    1. module docstring should be reformatted to follow coding standard
    1. `StringListMailbox` has no unit tests

I suspect there's more minor points and untested changes, but I'm exhausted.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun commented|
|-|-|

(In [25354]) address review points

except for `StringListMailbox` unit tests

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_30"></a>|@wsanchez commented|
|-|-|

Many of the review points are completely unrelated to this ticket and should be on a different branch, but whatever.  For example, missing unit tests for `StringListMailbox` has nothing to do with hashlib, and wasn't introduced by this fix.  While general code cleanup is nice and all, one fix shouldn't be blocked by unrelated fixes.

r25401 updates the copyright on every file changed on that branch as needed.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_31"></a>|@wsanchez commented|
|-|-|

[#3547](https://github.com/twisted/twisted/issues/3547) on `StringListMailbox`

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_32"></a>|@exarkun commented|
|-|-|

  1. Since Glyph merged [#10942](https://github.com/twisted/twisted/issues/10942), the `_fingerprint` function in twisted/conch/scripts/ckeygen.py can be replaced with `Key.fingerprint` (the branch needs to be merged forward for this).
  1. There's an extraneously added `raise` to the except suite in `printFingerprint` which should be removed.
  1. twisted/conch/test/test_ckeygen.py's copyright date should just say 2008, since it didn't exist in any of those prior years.
  1. in twisted/conch/test/test_ckeygen.py, there should be another blank line (for a total of two) between `KeyGenTests.setUp` and `KeyGenTests.test_printFingerprint`.
  1. in twisted/conch/ssh/transport.py, the commented out checks for the digest_size attribute are now doubly obsolete and should be deleted.
  1. twisted/news/database.py mixes up imports a bit and can lose the `nested_scopes` future import
  1. twisted/python/otp.py has no test coverage, lots of trailing whitespace, and other crummyness
  1. twisted/web/monitor.py has no test coverage; the module can't even be imported.
  1. The change to twisted/web2/server.py seems unrelated
  1. twisted/words/protocols/oscar.py should have its copyright date updated and the unused imports (time, nested_scopes) removed.
  1. twisted/words/test/test_msn.py has an unused import, `sys`, which should be removed

A bunch of these points are for code that I wrote after my last review anyway, so I'll merge the branch forward and address them.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_33"></a>|@exarkun commented|
|-|-|

(In [25426]) Branching to 'hashlib-2763-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_34"></a>|@exarkun commented|
|-|-|

(In [25428]) address points 1-5

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_35"></a>|@exarkun commented|
|-|-|

(In [25429]) address point 6

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_36"></a>|@exarkun commented|
|-|-|

(In [25430]) deprecate twisted.python.otp

I tried writing tests for it first, but it's either broken or not well
documented enough for me to figure out how to use.  In any case, there are
other OTP libraries for Python, and there is nothing specific or beneficial to
Twisted about this module.  It also hasn't really been touched since it was
added seven years ago.

No tests for this deprecation because I don't know how to test module level
warnings.

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_37"></a>|@exarkun commented|
|-|-|

(In [25431]) delete twisted.web.monitor (addresses review point 8)

This module has been unimportable for 4 years.  Someone noticed, and even cared
enough to file a ticket, but not enough to write any unit tests.  So, removing.
This isn't a backward incompatible change because the behavior is the same
before and after: the module cannot be imported.

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_38"></a>|@exarkun commented|
|-|-|

(In [25432]) cleanup imports, addresses review points 10 and 11

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_39"></a>|@exarkun set owner to @wsanchez<br>@exarkun set status to `new`|
|-|-|

Took care of all the points except for 9.  I'm not really sure what's up with that.

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)<a name="note_40"></a>|@wsanchez commented|
|-|-|

Reverted `twisted/web2/server.py`.  No clue how that change got onto that branch. `:-/`

**Comment by twisted-trac** [comment]
|[<img alt="wsanchez's avatar" src="https://avatars.githubusercontent.com/u/50002?s=50" width="50" height="50">](https://github.com/wsanchez)|@wsanchez set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_43"></a>|@exarkun removed owner|
|-|-|

Anybody can review this (and having someone other than me give it a look would be great).

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_44"></a>|@mwhudson set owner to @exarkun|
|-|-|

Mostly this is a monster yawn of course.  I have some small problems with the new tests you've written, and another comment:

 1. ```twisted.words.test.test_msn.NotificationTests.test_challenge```
    Reading this test, I have no idea how to tell if it is correct or not.  Given the nature of what this is testing, this is perhaps inevitable, but maybe you could compute the result rather than including it literally, and include a reference to why this is the right thing to do?

 2. ```twisted.test.test_newcred.DeprecatedUtilTests.test_respond```
    Same complaint really.  Also, I think it would result in clearer failures if you put the assertion about the length of warnings before the assertions about the contents of the list.

 3. Hasn't this branch missed the chance to be included in Twisted 8.2 by now?  If so, the deprecation messages need to be changed.

Finally, has this branch been tested on Python 2.6?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_45"></a>|@exarkun commented|
|-|-|

(In [25448]) Branching to 'hashlib-2763-3'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_46"></a>|@exarkun commented|
|-|-|

(In [25450]) try to clarify the assertions made in these hashy unit tests

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_47"></a>|@exarkun commented|
|-|-|

(In [25451]) change 8.2 deprecations to 8.3 deprecations

refs [#2763](https://github.com/twisted/twisted/issues/2763)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_48"></a>|@exarkun removed owner|
|-|-|

Points 1-3 addressed in the previous two commits.  The branch has been tested on Python 2.6:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/hashlib-2763 (search for py2.6).  The -3 branch is building now.  Results at:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/hashlib-2763-3

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)<a name="note_49"></a>|@mwhudson commented|
|-|-|

I'm still not completely convinced by test_respond, but I don't think I care.  Land it.

**Comment by twisted-trac** [comment]
|[<img alt="mwhudson's avatar" src="https://avatars.githubusercontent.com/u/672022?s=50" width="50" height="50">](https://github.com/mwhudson)|@mwhudson set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_51"></a>|@exarkun set status to `closed`|
|-|-|

(In [25457]) Merge hashlib-2763-3

Author: wsanchez, exarkun
Reviewer: exarkun, mwhudson
Fixes: [#2763](https://github.com/twisted/twisted/issues/2763)

Replace uses of `md5` and `sha` modules in Twisted with use of a new `twisted.python.hashlib`
module which transparently uses the new `hashlib` standard library module if it is available
or falls back to `md5` and `sha` if not.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
